File: julia/src/data_structures/stack.jl
"""
Stack - LIFO data structure

Time Complexity:
- push!: O(1) amortized
- pop!: O(1)
- peek: O(1)
- isempty: O(1)
- size: O(1)
"""

mutable struct Stack{T}
    items::Vector{T}

    Stack{T}() where T = new{T}(T[])
end

Stack() = Stack{Any}()

"""
    push!(s::Stack{T}, item::T) where T

Push an item onto the stack.
"""
function Base.push!(s::Stack{T}, item::T) where T
    push!(s.items, item)
    s
end

"""
    pop!(s::Stack{T}) where T

Remove and return the top item.
Throws ArgumentError if stack is empty.
"""
function Base.pop!(s::Stack{T}) where T
    if isempty(s)
        throw(ArgumentError("Stack is empty"))
    end
    pop!(s.items)
end

"""
    peek(s::Stack{T}) where T

Return the top item without removing it.
Throws ArgumentError if stack is empty.
"""
function peek(s::Stack{T}) where T
    if isempty(s)
        throw(ArgumentError("Stack is empty"))
    end
    s.items[end]
end

"""
    isempty(s::Stack)

Check if the stack is empty.
"""
Base.isempty(s::Stack) = isempty(s.items)

"""
    size(s::Stack)

Return the number of items in the stack.
"""
Base.size(s::Stack) = length(s.items)

"""
    length(s::Stack)

Return the number of items in the stack.
"""
Base.length(s::Stack) = length(s.items)

"""
    empty!(s::Stack)

Remove all items from the stack.
"""
function Base.empty!(s::Stack)
    empty!(s.items)
    s
end

"""
    collect(s::Stack)

Convert stack to array (bottom to top).
"""
Base.collect(s::Stack) = copy(s.items)

# Iterator interface
Base.iterate(s::Stack, state=1) = state > length(s.items) ? nothing : (s.items[state], state + 1)
Base.eltype(::Type{Stack{T}}) where T = T

File: lua/data_structures/stack.lua
-- Stack - LIFO data structure
--
-- Time Complexity:
-- - push: O(1)
-- - pop: O(1)
-- - peek: O(1)
-- - isEmpty: O(1)
-- - size: O(1)

local Stack = {}
Stack.__index = Stack

function Stack.new()
    local self = setmetatable({}, Stack)
    self.items = {}
    self.count = 0
    return self
end

function Stack:push(item)
    self.count = self.count + 1
    self.items[self.count] = item
end

function Stack:pop()
    if self.count == 0 then
        error("Stack is empty")
    end
    local item = self.items[self.count]
    self.items[self.count] = nil
    self.count = self.count - 1
    return item
end

function Stack:peek()
    if self.count == 0 then
        error("Stack is empty")
    end
    return self.items[self.count]
end

function Stack:isEmpty()
    return self.count == 0
end

function Stack:size()
    return self.count
end

function Stack:clear()
    self.items = {}
    self.count = 0
end

function Stack:toArray()
    local result = {}
    for i = 1, self.count do
        result[i] = self.items[i]
    end
    return result
end

return Stack
